### PR TITLE
Add command storeconfig

### PR DESCRIPTION
**Pull request by cocap10**
This PR adds the command `storeconfig` which stores the static traefik configuration into a Key-value stores.

TODO :  
- [x] code test-integration 
- [x] review code
- [x] review doc

### CODE REVIEW

File: docs/user-guide/kv-config.md [hunk 1]
```diff
@@ -55,7 +55,10 @@ whoami4:
 ## Upload the configuration in the Key-value store
 
 We should now fill the store with the Træfɪk global configuration, as we do with a [TOML file configuration](/toml).
-To do that, we can send the Key-value pairs via [curl commands](https://www.consul.io/intro/getting-started/kv.html) or via the [Web UI](https://www.consul.io/intro/getting-started/ui.html) 
+To do that, we can send the Key-value pairs via [curl commands](https://www.consul.io/intro/getting-started/kv.html) or via the [Web UI](https://www.consul.io/intro/getting-started/ui.html).
+
+Hopefully, we had automated this process with the `storeconfig` Træfɪk command. 
+Please refer to the [store Træfɪk configuration](/user-guide/kv-config/#store-configuration-in-key-value-store) section to get documentation on it.
 
 Here the toml configuration we would like to store in the Key-value Store  :
```
Comment (emilevauge) [inline docs/user-guide/kv-config.md]: Here is

**Comment by emilevauge** [comment]
You should specify that "In case you are setting key values manually", because you were talking about using `storeconfig` https://github.com/containous/traefik/pull/551/files?diff=unified#diff-91724a2dd65e9b63a22411496e3b9968R115

**Comment by emilevauge** [comment]
Not needed: https://github.com/containous/traefik/pull/551/files?diff=unified#diff-91724a2dd65e9b63a22411496e3b9968R137

**Comment by emilevauge** [comment]
Here is  again ;) https://github.com/containous/traefik/pull/551/files?diff=unified#diff-91724a2dd65e9b63a22411496e3b9968R172

File: docs/user-guide/kv-config.md [hunk 2]
```diff
@@ -295,4 +298,19 @@ Once the `/traefik/alias` key is updated, the new `/traefik_configurations/2` co
 | `/traefik_configurations/2/backends/backend1/servers/server2/weight`    | `5`                        |
 
 Note that Træfɪk *will not watch for key changes in the `/traefik_configurations` prefix*. It will only watch for changes in the `/traefik/alias`. 
-Further, if the `/traefik/alias` key is set, all other configuration with `/traefik/backends` or `/traefik/frontends` prefix are ignored.
\ No newline at end of file
+Further, if the `/traefik/alias` key is set, all other configuration with `/traefik/backends` or `/traefik/frontends` prefix are ignored.
+
+# Store configuration in Key-value store
+
+The static Træfɪk configuration can be 	automatically uploaded form Træfɪk into a Key-value Store.
```
Comment (emilevauge) [inline docs/user-guide/kv-config.md]: `uration can be 	automatically` -> soooo much spaces ^^

**Comment by emilevauge** [comment]
Well done @cocap10 :tada: 
I made few comments.

/cc @containous/traefik

**Comment by cocap10** [comment]
Thank you @emilevauge 
I modified where you commented ;)
Hope is ok now :)

**Comment by cocap10** [comment]
And rebased ;)

File: docs/user-guide/kv-config.md [hunk 3]
```diff
@@ -55,9 +55,12 @@ whoami4:
 ## Upload the configuration in the Key-value store
 
 We should now fill the store with the Træfɪk global configuration, as we do with a [TOML file configuration](/toml).
-To do that, we can send the Key-value pairs via [curl commands](https://www.consul.io/intro/getting-started/kv.html) or via the [Web UI](https://www.consul.io/intro/getting-started/ui.html) 
+To do that, we can send the Key-value pairs via [curl commands](https://www.consul.io/intro/getting-started/kv.html) or via the [Web UI](https://www.consul.io/intro/getting-started/ui.html).
 
-Here the toml configuration we would like to store in the Key-value Store  :
+Hopefully, we had automated this process with the `storeconfig` Træfɪk command.
```
Comment (vdemeester) [inline docs/user-guide/kv-config.md]: hum not sure about the *past* here.

```
Hopefully, Træfɪk allows automation of this process using the `storeconfig` subcommand.
```

File: docs/user-guide/kv-config.md [hunk 4]
```diff
@@ -295,4 +296,19 @@ Once the `/traefik/alias` key is updated, the new `/traefik_configurations/2` co
 | `/traefik_configurations/2/backends/backend1/servers/server2/weight`    | `5`                        |
 
 Note that Træfɪk *will not watch for key changes in the `/traefik_configurations` prefix*. It will only watch for changes in the `/traefik/alias`. 
-Further, if the `/traefik/alias` key is set, all other configuration with `/traefik/backends` or `/traefik/frontends` prefix are ignored.
\ No newline at end of file
+Further, if the `/traefik/alias` key is set, all other configuration with `/traefik/backends` or `/traefik/frontends` prefix are ignored.
+
+# Store configuration in Key-value store
+
+The static Træfɪk configuration can be automatically uploaded form Træfɪk into a Key-value Store.
```
Comment (vdemeester) [inline docs/user-guide/kv-config.md]: Maybe 

```
The static Træfɪk configuration in a key-value store can be automatically created and updated, using the [`storeconfig` subcommand](/basics/#commands).
```

(and remove the next line then 👼)

File: integration/consul_test.go [hunk 5]
```diff
@@ -392,3 +392,38 @@ func (s *ConsulSuite) skipTestGlobalConfigurationWithClientTLS(c *check.C) {
...
+		"/traefik/loglevel":                 "DEBUG",
+		"/traefik/defaultentrypoints/0":     "http",
+		"/traefik/entrypoints/http/address": ":8000",
+		"/traefik/web/address":              ":8080",
+		"/traefik/consul/endpoint":          (consulHost + ":8500"),
+	}
+
+	for key, value := range checkmap {
+		var p *store.KVPair
+		err = utils.Try(60*time.Second, func() error {
```
Comment (vdemeester) [inline integration/consul_test.go]: 60 seconds might be a little too much to wait for 👼

File: traefik.go [hunk 6]
```diff
@@ -84,6 +84,28 @@ Complete documentation is available at https://traefik.io`,
 		},
 	}
 
+	//storeconfig Command init
```
Comment (vdemeester) [inline traefik.go]: Note for later: we might want to split `traefik.go` a little bit 👼
(having a file for `storeconfig` command, same for `version`, …)

File: traefik.go [hunk 7]
```diff
@@ -114,13 +137,19 @@ Complete documentation is available at https://traefik.io`,
...
 		os.Exit(-1)
 	}
 
-	if kv != nil {
+	usedCmd, err := f.GetCommand()
+	if err != nil {
+		fmtlog.Println(err)
+		os.Exit(-1)
+	}
+	// IF a KV Store is enable and no sub-command called in agrs
```
Comment (vdemeester) [inline traefik.go]: s/agrs/args/

File: traefik.go [hunk 8]
```diff
@@ -114,13 +137,19 @@ Complete documentation is available at https://traefik.io`,
...
+	if kv != nil && usedCmd == traefikCmd {
```
Comment (vdemeester) [inline traefik.go]: This feels a little weird, to have to check that no sub-command have been called.
Naively I would think that only the code specified in the sub-command would be called. Might need some refactoring in the future 👼

**Comment by vdemeester** [comment]
I guess it's for 1.1, so I added the milestone.

@cocap10 few nits/comments but overall looks pretty good !

File: traefik.go (hunk 8 above)
Comment (cocap10) [inline traefik.go]: If the sub-command `storeconfig` is called, we want to parse all sources (flæg and toml) except the KV source (prevent update issues). 
Maybe checking the called command is not beautiful; but I d'ont see any other way so far :sweat_smile:

File: traefik.go (hunk 7 above)
Comment (cocap10) [inline traefik.go]: > s/agrs/args/

Ok, `agrs` -> `args` ;)
any other spelling mistake here ?
(I don't get the "s/" :angel: )

Comment (vdemeester) [inline traefik.go]: it's the `vim` command to say "replace this with this" 👼
(I know I use `emacs` but :trollface:)

File: traefik.go [hunk 9]
```diff
@@ -114,13 +137,19 @@ Complete documentation is available at https://traefik.io`,
...
 		s.AddSource(kv)
 		if _, err := s.LoadConfig(); err != nil {
 			fmtlog.Println(err)
```
Comment (emilevauge) [inline traefik.go]: No exit here ?

File: traefik.go (hunk 6 above)
Comment (cocap10) [inline traefik.go]: You would like to add a file "commands.go" ?
With commands definition inside ?

File: integration/consul_test.go (hunk 5 above)
Comment (cocap10) [inline integration/consul_test.go]: I copied it like that form "TestNominalConfiguration" :sweat_smile: 
How long do you think ?

File: traefik.go (hunk 9 above)
Comment (cocap10) [inline traefik.go]: You'r right ;)
And same there https://github.com/containous/traefik/pull/551/files#diff-dba3f2b16983e1351d5b02a5700d18ecR134 I think !

File: integration/consul_test.go (hunk 5 above)
Comment (vdemeester) [inline integration/consul_test.go]: I would say 15 seconds should be enough 👼.

File: traefik.go (hunk 6 above)
Comment (vdemeester) [inline traefik.go]: Not sure yet 👼

**Comment by emilevauge** [comment]
LGTM

**Comment by vdemeester** [comment]
LGTM 🐸

**Comment by Russell-IO** [comment]
LGTM

**Pull request merged**